### PR TITLE
Example: JupyterLab with Kernel Provider on Kubernetes

### PR DESCRIPTION
**Pull request by echarles**
Adds an example for JupyterLab with Kernel Provider on Kubernetes

### CODE REVIEW

**Review by kevin-bates — COMMENTED** [review]
Thanks for putting this together.  I've given some initial comments.  Lots of changes will be likely in the gateway-experiments repos.

File: examples/jlab-kernel-provider-k8s/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,93 @@
+# FROM jupyter/scipy-notebook:04f7f60d34a6
+FROM elyra/kernel-spark-py:2.0.0rc2
```
[the remark this answers is not in the document]
Comment (echarles) [reply examples/jlab-kernel-provider-k8s/Dockerfile]: Thx. have updated. PS: for now I have committed python-kernel only, prolly spark will be a next step.

File: examples/jlab-kernel-provider-k8s/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+#    jupyter k8s-kernelspec install --language=R && \
+#    jupyter k8s-kernelspec install --language=R --spark && \
+#    jupyter k8s-kernelspec install --language=Scala && \
+#    jupyter k8s-kernelspec install --language=Scala --spark
+
+RUN cd && \
+  cd enterprise_gateway && \
+  mkdir -p /usr/local/share/jupyter/kernels && \
+  cp -r etc/kernelspecs/spark_python_kubernetes /usr/local/share/jupyter/kernels/ && \
+  cp -r etc/kernel-launchers/kubernetes/scripts /usr/local/share/jupyter/kernels/spark_python_kubernetes/
```
[the remark this answers is not in the document]
Comment (echarles) [reply examples/jlab-kernel-provider-k8s/Dockerfile]: I have added that comment. Let's iterate on that. We need have a plan. for the repos... to be discussed at next server meeting.

File: examples/jlab-kernel-provider-k8s/jlab.yaml [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+        nb-selector: jupyterlab
+        app: jupyterlab
+        component: jupyterlab
+    spec:
+      # Created above.
+      serviceAccountName: jupyterlab-sa
+      containers:
+      - env:
+        - name: EG_PORT
+          value: "8888"
```
[the remark this answers is not in the document]
Comment (echarles) [reply examples/jlab-kernel-provider-k8s/jlab.yaml:114]: Have added that as comment.

File: examples/jlab-kernel-provider-k8s/jlab.yaml [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+          value: "3600"
+
+        - name: EG_LOG_LEVEL
+          value: "DEBUG"
+
+        - name: EG_KERNEL_LAUNCH_TIMEOUT
+          value: "60"
+
+        - name: EG_KERNEL_WHITELIST
+          value: "['r_kubernetes','python_kubernetes','python_tf_kubernetes','python_tf_gpu_kubernetes','scala_kubernetes','spark_r_kubernetes','spark_python_kubernetes','spark_scala_kubernetes']"
```
[the remark this answers is not in the document]
Comment (echarles) [reply examples/jlab-kernel-provider-k8s/jlab.yaml:140]: We will iterate on that. Have added it as comment for now.

**Comment by echarles** [comment]
@kevin-bates Thx a lot for your comments and the debugging session we have done yesterday.

I have invested more time today to debug the launch and the creation steps of the remote kernel. Something I have realized is that the python source changes are not picked up after the first invocation of those sources (even removing the `pyc` files). I have finally been able to have the pod remaining. Tried the grace-period... but finally simply remove the code responsible to remove the resource. The pod was working fine... so I have a second look at https://github.com/gateway-experiments/remote_kernel_provider/commit/fc0fdeaa0dbfe057e0fb4fd28fa4589713004a9b and it turns out that this was the cause.

Current commit works fine

Screenshot 2020-09-04 at 12 16 48

Let's further discuss the next steps at the weekly meeting.

I may add spark to this, but this is ready for review.

File: examples/jlab-kernel-provider-k8s/Dockerfile [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+# RUN cd && \
+#   export IR=https://github.com/gateway-experiments/remote_kernel_provider/releases/download/tag/v0.3.0 && \
+#   mkdir /usr/local/share/jupyter/kernels && \
+#   cd  /usr/local/share/jupyter/kernels && \
+#   curl $IR/yarn_kernelspecs.tar | tar xvf -
+
+RUN cd && \
+  git clone https://github.com/jupyter/enterprise_gateway && \
+  cd enterprise_gateway && \
+  pip install -e .
```
[the remark this answers is not in the document]
Comment (echarles) [reply examples/jlab-kernel-provider-k8s/Dockerfile:41]: Thx. Have removed unneeded commands

File: examples/jlab-kernel-provider-k8s/jlab.yaml [hunk 6]
```diff
@@ -0,0 +1,194 @@
...
+    resources: ["rolebindings"]
+    verbs: ["get", "list", "create", "delete"]
+---
+apiVersion: rbac.authorization.k8s.io/v1beta1
+kind: ClusterRole
+metadata:
+  # Referenced by EG_KERNEL_CLUSTER_ROLE below
+  name: kernel-controller
+  labels:
+    app: enterprise-gateway
```
[the remark this answers is not in the document]
Comment (echarles) [reply examples/jlab-kernel-provider-k8s/jlab.yaml]: Thx. This is changed.

**Comment by echarles** [comment]
Closing as https://github.com/jupyter/enhancement-proposals/pull/45 has not been merged.

**Pull request closed without merging**